jsprit-core/src/main/java/com/graphhopper/jsprit/core/util/ActivityTimeTracker.java

```diff
@@ -85,22 +85,39 @@ public void visit(TourActivity activity) {
         double transportTime = this.transportTime.getTransportTime(prevAct.getLocation(), activity.getLocation(), startAtPrevAct, route.getDriver(), route.getVehicle());
         double arrivalTimeAtCurrAct = startAtPrevAct + transportTime;
 
-        actArrTime = arrivalTimeAtCurrAct;
+        // modify the activity arrival time if this activity can be grouped with the previous one
+        // they will both have the same arrival and end times afterwards
+        if (canGroupActivities(activity, arrivalTimeAtCurrAct)) {
+            actArrTime = arrivalTimeAtCurrAct - prevAct.getOperationTime();
+        } else {
+            actArrTime = arrivalTimeAtCurrAct;
+        }
         double operationStartTime;
 
         if (activityPolicy.equals(ActivityPolicy.AS_SOON_AS_TIME_WINDOW_OPENS)) {
             operationStartTime = Math.max(activity.getTheoreticalEarliestOperationStartTime(), arrivalTimeAtCurrAct);
         } else if (activityPolicy.equals(ActivityPolicy.AS_SOON_AS_ARRIVED)) {
             operationStartTime = actArrTime;
+        } else if (activityPolicy.equals(ActivityPolicy.AS_SOON_AS_TIME_WINDOW_OPENS_WITHIN_GROUP)) {
+            operationStartTime = Math.max(activity.getTheoreticalEarliestOperationStartTime(), actArrTime);
         } else operationStartTime = actArrTime;
 
-        double operationEndTime = operationStartTime + activityCosts.getActivityDuration(activity,actArrTime,route.getDriver(),route.getVehicle());
+        double operationEndTime;
+        // if the current activity can be grouped with the previous one adjust the operation end time
+        // select the operation time which is bigger
+        // as we iterate over each activity, we need to change the operation end time of the previous activity so that they have the same end time
+        // (we didn't know when inserting the previous activity, that we should use the operating time of the current activity)
+        if (canGroupActivities(activity, arrivalTimeAtCurrAct)) {
+            operationEndTime = operationStartTime + Math.max(prevAct.getOperationTime(), activity.getOperationTime());
+            prevAct.setEndTime(operationEndTime);
```

Comment:
This should probably move to `UpdateActivityTimes` class, so that `UpdateVariableCosts` class is unaffected